src/pextlib1.0/Pextlib.c

```diff
@@ -648,6 +651,424 @@ int SetMaxOpenFilesCmd(ClientData clientData UNUSED, Tcl_Interp *interp, int obj
 	return TCL_OK;
 }
 
+/* Mount point file system case-sensitivity caching infrastructure. */
+typedef struct _mount_cs_cache_entry {
+    char *mountpoint;
+    int case_sensitive;
+} mount_cs_cache_entry_t, *mount_cs_cache_entry_list_t;
+
+struct _mount_cs_cache {
+    size_t count;
+    mount_cs_cache_entry_t **entries;
+};
+
+/**
+ * Returns a new pre-allocated mount_cs_cache_t object.
+ */
+mount_cs_cache_t* new_mount_cs_cache() {
+    mount_cs_cache_t *ret = malloc(sizeof(mount_cs_cache_t));
+
+    if (ret) {
+        ret->count = 0;
+        ret->entries = NULL;
+    }
+
+    return ret;
+}
+
+/**
+ * Resets a mount cache object.
+ */
+void reset_mount_cs_cache(mount_cs_cache_t *cache) {
+    if (cache) {
+        /*
+         * Assume that if the count of cached entries is zero,
+         * the entries handle is set to NULL as well. Likewise,
+         * if the count is non-zero, the entries pointer should
+         * be non-NULL. Any other combination probably means
+         * something is very, very wrong.
+         */
+        assert((!cache->count && !cache->entries) || (cache->count && cache->entries));
+
+        for (size_t i = 0; i < cache->count; ++i) {
+            free(cache->entries[i]->mountpoint);
+            cache->entries[i]->mountpoint = NULL;
+
+            free(cache->entries[i]);
+            cache->entries[i] = NULL;
+        }
+
+        free(cache->entries);
+        cache->entries = NULL;
+    }
+}
+
+/**
+ * Rollback mount cache object.
+ */
+static void rollback_mount_cs_cache(Tcl_Interp *interp, mount_cs_cache_t *cache) {
```

Comment:
I wouldn't have gone the extra mile to shrink this again on failed allocations. The bit of extra unused memory is not a problem (it's neither lost nor unreachable, nor will it be used uninitialized) when `cache->count` isn't increased.

Now that you have it, let's keep it.